internal/restic/counted_blob_set.go

```diff
@@ -0,0 +1,66 @@
+package restic
+
+import "sort"
+
+// CountedBlobSet is a set of blobs.
+type CountedBlobSet map[BlobHandle]uint8
+
+// NewCountedBlobSet returns a new CountedBlobSet, populated with ids.
+func NewCountedBlobSet(handles ...BlobHandle) CountedBlobSet {
+	m := make(CountedBlobSet)
+	for _, h := range handles {
+		m[h] = 0
+	}
+
+	return m
+}
+
+// Has returns true iff id is contained in the set.
+func (s CountedBlobSet) Has(h BlobHandle) bool {
+	_, ok := s[h]
+	return ok
+}
+
+// Insert adds id to the set.
+func (s CountedBlobSet) Insert(h BlobHandle) {
+	s[h] = 0
+}
+
+// Delete removes id from the set.
+func (s CountedBlobSet) Delete(h BlobHandle) {
+	delete(s, h)
+}
+
+func (s CountedBlobSet) Len() int {
+	return len(s)
+}
+
+// List returns a sorted slice of all BlobHandle in the set.
+func (s CountedBlobSet) List() BlobHandles {
+	list := make(BlobHandles, 0, len(s))
+	for h := range s {
+		list = append(list, h)
+	}
+
+	sort.Sort(list)
+
+	return list
+}
+
+func (s CountedBlobSet) String() string {
+	str := s.List().String()
+	if len(str) < 2 {
+		return "{}"
+	}
+
+	return "{" + str[1:len(str)-1] + "}"
+}
+
+// Copy returns a copy of the CountedBlobSet.
+func (s CountedBlobSet) Copy() CountedBlobSet {
+	cp := NewCountedBlobSet()
```

Comment:
I didn't test this, but I think a size hint should produce an even smaller map:

```suggestion
	cp := make(CountedBlobSet, len(s))
```

Reply:
The bucket array apparently always uses a power of 2 as size. But providing a hint should reduce the necessary reallocations. In particular we shouldn't end up with a half complete map growth operations that has to keep the old and new bucket array around.

https://github.com/golang/go/blob/4e3d805a015153ca4e3382415c2474e4f954ddee/src/runtime/map.go#L318